srml/transaction-payment/rpc/src/lib.rs

```diff
@@ -0,0 +1,105 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! RPC interface for the t
+//! ransa
+//!ction payment module.
+use std::sync::Arc;
+use serde::{Serialize, Deserialize};
+use codec::Codec;
+use client::blockchain::HeaderBackend;
+use jsonrpc_core::{Error, ErrorCode, Result};
+use jsonrpc_derive::rpc;
+use sr_primitives::{
+	generic::BlockId,
+	traits::{Block as BlockT, ProvideRuntimeApi},
+};
+use srml_transaction_payment_rpc_runtime_api::{
+	TransactionPaymentApi as TransactionPaymentRuntimeApi, RuntimeDispatchInfo,
+};
+use self::gen_client::Client as TransactionPaymentClient;
+
+
+/// A wrapper around the request information to query a dispatchable info.
+#[derive(Serialize, Deserialize)]
+pub struct DispatchInfoRequest<AccountId, Extra, Call> {
+	/// Sender aka. origin of the dispatch.
+	pub origin: AccountId,
+	/// The call type.
+	pub call: Call,
+	/// Signed extensions.
+	pub extra: Extra,
+}
+
+#[rpc]
+pub trait TransactionPaymentApi<BlockHash, AccountId, Balance, Call, Extra> {
+	#[rpc(name = "transaction_payment_query_unsigned")]
+	fn query_info_unsigned(
+		&self,
+		request: DispatchInfoRequest<AccountId, Extra, Call>
+	) -> Result<RuntimeDispatchInfo<Balance>>;
+}
+
+/// A struct that implements the [`TransactionPaymentApi`].
+pub struct TransactionPayment<C, B> {
+	client: Arc<C>,
+	_marker: std::marker::PhantomData<B>,
+}
+
+impl<C, B> TransactionPayment<C, B> {
+	pub fn new(client: Arc<C>) -> Self {
+		TransactionPayment { client, _marker: Default::default() }
+	}
+}
+
+impl<
+	C,
+	Block,
+	AccountId,
+	Balance,
+	Call,
+	Extra,
+> TransactionPaymentApi<
+	<Block as BlockT>::Hash,
+	AccountId,
+	Balance,
+	Call,
+	Extra,
+> for TransactionPayment<C, Block> where
+	Block: BlockT,
+	C: Send + Sync + 'static + ProvideRuntimeApi + HeaderBackend<Block>,
+	C::Api: TransactionPaymentRuntimeApi<Block, AccountId, Balance, Call, Extra>,
+	AccountId: Codec,
+	Balance: Codec,
+	Call: Codec,
+	Extra: Codec,
+{
+	fn query_info_unsigned(
+		&self,
+		request: DispatchInfoRequest<AccountId, Extra, Call>
+	) -> Result<RuntimeDispatchInfo<Balance>> {
+		let api = self.client.runtime_api();
+		let best = self.client.info().best_hash;
```

Comment:
The convention should be to allow specifying the hash as (optional) last argument and only fall back to `best_hash` if it's not provided.

Reply:
I am not sure if this really makes any sense here, I only followed examples after realizing that apparently all of them require the best block as the argument. 

How is this used by the runtime api? if the given block is not the best, the runtime call is made to the runtime state at the given particular block? this sounds quite expensive to do actually and I don't know if it makes sense for anyone to ask for previous fees?

Reply:
Yes, we call into the state of a specific block if the hash is passed and that state is not pruned yet.
It should be as expensive as calling into latest state block unless I'm missing something obvious here.

The reason for that is that you might have multiple calls that you want to do in parallel (or sequentially) and would be good to enforce that they are run against the same state.
So pretty much you fetch the latest hash yourself and then spawn a bunch of calls and pass that hash. Now you know that the estimates are coming from the same latest block and not from potentially different states. To be fair I can't yet imagine why this would be useful for estimating the tip, but would prefer to maintain consistency between RPCs here.

Reply:
Yes my main concern was that what if someone (just because they can it is fun) decides to query the fee price of some old past. 

This brings an interesting thought: the runtime code itself is also part of the state right (`:code`)? given this, if there is a function `foo()` which is even removed from the runtime, I can still query it by given a block hash at which this function existed in the runtime.  

Thanks for the explanation. Will add the optional block hash.

Reply:
> if there is a function foo() which is even removed from the runtime, I can still query it by given a block hash at which this function existed in the runtime.

Yes, but that's also why we have `RuntimeApi`s, so that we can query what functions are available or change the implementation without affecting the API.